### PR TITLE
update registerhandling to use i1* pointer

### PR DESCRIPTION
**Pull request by cqc-melf**

### CODE REVIEW

**Comment by cqc-melf** [comment]
The discussed changes for the classical register handling

File: pytket/qir/conversion/conversion.py [hunk 1]
```diff
@@ -129,35 +130,35 @@ def __init__(
         self.set_cregs: dict[str, list] = {}  # Keep track of set registers.
         self.ssa_vars: dict[str, Value] = {}  # Keep track of set ssa variables.
 
-        # i1 read_bit_from_reg(i64 reg, i64 index)
+        # i1 read_bit_from_reg(i1* reg, i64 index)
```
Comment (qartik) [inline pytket/qir/conversion/conversion.py]:
```suggestion
        # i1 get_creg_bit(i1* creg, i64 index)
```

File: pytket/qir/conversion/conversion.py [hunk 2]
```diff
@@ -129,35 +130,35 @@ def __init__(
...
             "read_bit_from_reg",
             pyqir.FunctionType(
                 pyqir.IntType(self.module.module.context, 1),
-                [self.qir_int_type] * 2,
+                [self.qir_i1p_type, self.qir_int_type],
             ),
         )
 
-        # void set_one_bit_in_reg(i64 reg, i64 index, i1 value)
+        # void set_one_bit_in_reg(i1* reg, i64 index, i1 value)
```
Comment (qartik) [inline pytket/qir/conversion/conversion.py]:
```suggestion
        # void set_creg_bit(i1* creg, i64 index, i1 value)
```

File: pytket/qir/conversion/conversion.py [hunk 3]
```diff
@@ -129,35 +130,35 @@ def __init__(
...
-                    self.qir_int_type,
+                    self.qir_i1p_type,
                     self.qir_int_type,
                     pyqir.IntType(self.module.module.context, 1),
                 ],
             ),
         )
 
-        # void set_all_bits_in_reg(i64 reg, i64 value)
+        # void set_all_bits_in_reg(i1* reg, i64 value)
```
Comment (qartik) [inline pytket/qir/conversion/conversion.py]:
```suggestion
        # void set_creg_to_int(i1* creg, i64 value)
```

File: pytket/qir/conversion/conversion.py [hunk 4]
```diff
@@ -172,12 +173,23 @@ def __init__(
             ),
         )
 
-        # i64 reg2var(i1, i1, i1, ...)
-        self.reg2var = self.module.module.add_external_function(
-            "reg2var",
+        # i1* create_reg(i64 size)
```
Comment (qartik) [inline pytket/qir/conversion/conversion.py]:
```suggestion
        # i1* create_creg(i64 size)
```

File: pytket/qir/conversion/conversion.py [hunk 5]
```diff
@@ -172,12 +173,23 @@ def __init__(
...
+            "create_reg",
             pyqir.FunctionType(
-                pyqir.IntType(self.module.module.context, qir_int_type),
-                [pyqir.IntType(self.module.module.context, 1)] * qir_int_type,
+                self.qir_i1p_type,
+                [pyqir.IntType(self.module.module.context, qir_int_type)],
+            ),
+        )
+
+        # i64 create_reg(i1* reg)
```
Comment (qartik) [inline pytket/qir/conversion/conversion.py]:
```suggestion
        # i64 get_int_from_creg(i1* creg)
```
Specifically `read_all_bits_from_reg` could be better called `get_int_from_creg`

**Review by qartik — CHANGES_REQUESTED** [review]
I think we talked about settling on better names for the API functions. I have left some suggestions. Perhaps it makes sense to list them out all together:

```llvm
i1*  create_creg      (i64 size                     )
void set_creg_to_int  (i1* creg, i64 value          )
i64  get_int_from_creg(i1* creg                     )
void set_creg_bit     (i1* creg, i64 index, i1 value)
i1   get_creg_bit     (i1* creg, i64 index          )
```

Note that in the backend compiler, a notion of `creg` is prevalent so perhaps it makes sense to also use it for this API.

I am open to hear about other suggestions. cc: @peter-campora

**Comment by cqc-melf** [comment]
> I think we talked about settling on better names for the API functions. I have left some suggestions. Perhaps it makes sense to list them out all together:
> 
> ```llvm
> i1*  create_creg      (i64 size                     )
> void set_creg_to_int  (i1* creg, i64 value          )
> i64  get_int_from_creg(i1* creg                     )
> void set_creg_bit     (i1* creg, i64 index, i1 value)
> i1   get_creg_bit     (i1* creg, i64 index          )
> ```
> 
> Note that in the backend compiler, a notion of `creg` is prevalent so perhaps it makes sense to also use it for this API.
> 
> I am open to hear about other suggestions. cc: @peter-campora

I have done a search and replace as suggested, I have replaced that in all of the ll files for the tests as well.

See @cqc-melf
[rename functions as suggested](https://github.com/CQCL/pytket-qir/pull/61/commits/867e2a8f511d225eee6a99b4efcbb942a1c69897)
[867e2a8](https://github.com/CQCL/pytket-qir/pull/61/commits/867e2a8f511d225eee6a99b4efcbb942a1c69897)

And @cqc-melf
[rename reg to creg in comments](https://github.com/CQCL/pytket-qir/pull/61/commits/3a4cfdd32210724451ec9f7b5e4dabb7a6ee64b8)
[3a4cfdd](https://github.com/CQCL/pytket-qir/pull/61/commits/3a4cfdd32210724451ec9f7b5e4dabb7a6ee64b8)

**Review by cqc-alec — APPROVED** [review]

**Review by qartik — APPROVED** [review]
Thanks for the careful changes!

**Pull request merged**